### PR TITLE
Core: Allow manifest file cache to be configurable

### LINKED ISSUE #9991: Improvements for manifest file caching
### Feature Request / Improvement

While experimenting with the features in #4518 (core: Provide mechanism to cache manifest file content), we encountered a couple of limitations which we would like to improve on and wanted to see if there are any thoughts or feedback from the Iceberg community.

### Add support for modifying the cache key

Manifest caching assumes that the key to the manifest cache is a single FileIO instance.  This presumption seems difficult in certain circumstances, for example, when authenticating with Hadoop and you are using impersonation for authentication—-it’s unclear how to do this with the reference implementation without creating multiple FileIOs per user, which would degrade the cache hit rate.

To allow for this change in caching behavior, I am wondering if it makes sense to allow FileIO instances to supply a cache key.  This way, users of the Iceberg reference implementation can supply their own FileIO (as I believe is already common practice among some query engines), and in this implementation you could specify a cache key that achieves better hit rate across multiple FileIO instances.

### Allow users to retrieve the cache statistics/potentially supply their own Caffeine cache

Ideally, when we set the cache, operators would have a way of monitoring the cache hit rate for effectiveness (to see if the cache needs to be tuned), or to simply supply a cache of their own, of which operators can then set up independent monitoring.  Right now, I believe the only way to access the stats is through debug-level logging, and there is not public interface exposed to retrieve the metrics.

I haven’t found a way to do this yet, but I think if possible, it would be a really great idea to allow users to supply their own cache, with their own caching parameters, which they can then set up observability over so they can export these cache metrics to e.g. Prometheus.  In lieu of that though, I think a simple solution would be to simply allow the cache statistics to be exported from ManifestFiles.

---

I wanted to get some feedback on whether or not the community would be supportive of these contributions, or had feedback or different ideas on how to go about this.

### Query engine

PrestoDB

### PR DESCRIPTION
**Pull request by tdcmeehan**
Refactor manifest file caching logic in ManifestFiles class to split out cache management and allow for customizable cache manager.

This is one approach to fix #9991.  By allowing users to inject their own cache management, users can monitor cache metrics independently, influence caching behavior, and provide application-specific caching that is currently difficult using the built in cache manager.

Note that `DefaultContentCache` is renamed from `ContentCache`, and the code is mostly the same.  `ContentCach` is a new interface derived from `ContentCache`.  Most of the code in `ContentCacheManager` is derived from `ManifestFiles` and is meant to allow the logic inside `ManifestFiles` to be customized or overridden.

### CODE REVIEW

**Comment by singhpk234** [comment]
interesting, whats the use case for this ?

**Comment by tdcmeehan** [comment]
Thanks for your question @singhpk234.  The motivation is to try to make both improvements listed in #9991, and I would like to solicit feedback on this approach.

tl;dr, there are two problems I'd like to solve: I would like for manifest caching to expose metrics to our project's existing metrics reporting infrastructure, and I would also like for manifest caching to be able to have a customizable cache key to account for custom implementations of FileIO which don't make the same presumptions as the default manifest cache, namely that there is single long-lived instance of each type of FileIO.

While there are other approaches to solve both of these problems, it seemed most straightforward to allow the manifest cache to be pluggable--that way, I can easily integrate it with my existing infrastructure, as I'm in control of the code which supplies the cache.  And by allowing me to write a custom cache, I am in complete control of my project's caching needs, without having to make a lot of changes in the core library.

The feedback I would request is: does the Iceberg core library want these things as well?  i.e., does the core library want the ability to handle custom FileIO caching which does not use long-lived FileIO references, and does the core project want to expose metrics for manifest caching?  In order to support such things cleanly, we may require some refactoring to FileIO and ManifestCaches.  I have thoughts around that, but wanted to get feedback on the community before I proceed either way.  (My personal thinking is, caching is a bit of an involved infrastructure investment, and it may make sense to simply provide a basic implementation (as done here), but allow arbitrary cache customization for those that require such sophistication, rather than build it into the core library.)

File: core/src/main/java/org/apache/iceberg/io/DefaultContentCache.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+ * <p>The file-content caching is initiated by calling {@link DefaultContentCache#tryCache(InputFile)}.
+ * Given a FileIO, a file location string, and file length that is within allowed limit,
+ * DefaultContentCache will return a {@link CachingInputFile} that is backed by the cache. Calling {@link
+ * CachingInputFile#newStream()} will return a {@link ByteBufferInputStream} backed by list of
+ * {@link ByteBuffer} from the cache if such file-content exist in the cache. If the file-content
+ * does not exist in the cache yet, a regular InputFile will be instantiated, read-ahead, and loaded
+ * into the cache before returning ByteBufferInputStream. The regular InputFile is also used as a
+ * fallback if cache loading fail.
+ */
+public class DefaultContentCache implements ContentCache {
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/io/DefaultContentCache.java:51]: `InMemoryContentCache`? I understand it's the default impl but feels a bit more clear to have InMemory be in the class name since it indicates the behavior.

File: core/src/main/java/org/apache/iceberg/io/ContentCache.java [hunk 2]
```diff
@@ -18,274 +18,35 @@
...
  * Given a FileIO, a file location string, and file length that is within allowed limit,
- * ContentCache will return a {@link CachingInputFile} that is backed by the cache. Calling {@link
- * CachingInputFile#newStream()} will return a {@link ByteBufferInputStream} backed by list of
- * {@link ByteBuffer} from the cache if such file-content exist in the cache. If the file-content
- * does not exist in the cache yet, a regular InputFile will be instantiated, read-ahead, and loaded
- * into the cache before returning ByteBufferInputStream. The regular InputFile is also used as a
- * fallback if cache loading fail.
+ * ContentCache will return an implementation of a {@link InputFile} that may be cached.
  */
-public class ContentCache {
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/io/ContentCache.java:51]: If we go with such an approach we'll need to be mindful of backwards compatibility since this is public; even though this is now an interface others can currently be doing a new ContentCache(....) and so a direct removal would break those.
We'd need to put this on a deprecation path

File: core/src/main/java/org/apache/iceberg/io/ContentCacheManager.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+/**
+ * Manages content caches for FileIO implementations, capable of creating or retrieving
+ * a {@link ContentCache} given a {@link FileIO}, or dropping a cache for a {@link FileIO}.
+ */
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/io/ContentCacheManager.java:25]: Sorry not sure if I completely follow, why is `ContentCacheManager` required? Couldn't we achieve the desired goals with just having a pluggable `ContentCache`?

**Review by amogh-jahagirdar — COMMENTED** [review]
@tdcmeehan I think having a pluggable cache makes sense. People may want to have their own caching tiers or be able to collect metrics in your case, etc. It falls in line with general principles followed in Iceberg (pluggable catalog, FileIO, etc). 

In particular with the proposed changes I had 2 main comments:

1.) We need to be mindful of compatibility, so we can't just remove public classes. If there's consensus on this, we'd need to have a deprecation path for the relevant classes.

2.) I don't really follow why we need a separate pluggable `ContentCacheManager` ? Why isn't a separate pluggable `ContentCache` sufficient?

File: core/src/main/java/org/apache/iceberg/io/ContentCacheManager.java (hunk 3 above)
Comment (tdcmeehan) [reply core/src/main/java/org/apache/iceberg/io/ContentCacheManager.java:25]: The reason for additionally adding a `ContentCacheManager`, is to override the cache key.  The current default in `ManifestFiles` presumes a single long lived instance of a `FileIO`.  However, users of custom FileIOs may not make this presumption.  In order to accommodate that, you can override the `ContentCacheManager`.  The alternative might be to specify a custom key in the `FileIO` itself.  Additionally, by introducing a manager, it seemed easier to provide global metrics of cache utilization across FileIO instances.

File: core/src/main/java/org/apache/iceberg/io/ContentCache.java (hunk 2 above)
Comment (tdcmeehan) [reply core/src/main/java/org/apache/iceberg/io/ContentCache.java:51]: I have marked the class as deprecated according to the deprecation notice: https://iceberg.apache.org/contribute/#deprecation-notices

I have left this class just about unchanged, and copied a class `InMemoryContentCache`.  This class is marked as deprecated and the documentation points to the in memory cache.  The interface name is changed to `FileIOContentCache`.  PTAL and let me know what you think of this revision.

**Comment by tdcmeehan** [comment]
Hello @singhpk234 @amogh-jahagirdar thanks for your earlier reviews.  I see this PR has stalled--is there any way to move it forward?

Once again, the intention is to make manifest file caching more generally useful for other users of the Iceberg library by making it configurable.  Please let me know if you have any suggestions or feedback on this approach.

**Comment by yingsu00** [comment]
Gentle ping @singhpk234 @amogh-jahagirdar thanks!

**Comment by shenh062326** [comment]
We also encountered a similar problem. It would be helpful if this MR could be merged into iceberg.
